File: 0322-coin-change/0322-coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        if amount == 0:
            return 0
        q, seen = deque([(0, 0)]), set()
        while q:
            num_coins, val = q.popleft()
            num_coins += 1
            for coin in coins:
                next_val = val + coin
                if next_val == amount:
                    return num_coins
                if next_val < amount:
                    if next_val not in seen:
                        seen.add(next_val)
                        q.append((num_coins, next_val))
        return -1
        
        #
        
        dp = [inf for _ in range(0, amount+1)]
        dp[0] = 0
        for i in range(1, amount+1):
            for coin in coins:
                if coin <= i:
                    dp[i] = min(dp[i], 1 + dp[i-coin])
        return -1 if dp[amount] == inf else dp[amount]

File: 0322-coin-change/0322-coin-change.ts
function coinChange(coins: number[], amount: number): number {
    const dp: number[] = Array(amount+1).fill(Infinity)
    dp[0] = 0
    for (let amt=1; amt <= amount; amt++) {
        for (const coin of coins) {
            if (coin <= amt) {
                dp[amt] = Math.min(dp[amt], 1 + dp[amt-coin])
            }
        }
    }
    return dp[amount] === Infinity ? -1 : dp[amount]
    
//     const dp = {}
//     const min = helper(amount)
//     return min === Infinity ? -1 : min
    
//     function helper(amount: number): number {
//         if (amount === 0) return 0
//         if (amount in dp) return dp[amount]
//         let res = Infinity
//         for (const coin of coins) {
//             if (coin <= amount) {
//                 res = Math.min(res, 1 + helper(amount-coin))
//             }
//         }
//         return dp[amount] = res
//     }
};